src/detect-metadata.c

```diff
@@ -28,23 +28,271 @@
+        if (res < 0) {
+            SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_copy_substring failed");
+            goto error;
+        }
+        key = DetectMedatataHashAdd(de_ctx, pkey);
+        if (key == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "can't create metadata key");
+            goto error;
+        }
+
+        res =  pcre_copy_substring(kv, ov, MAX_SUBSTRINGS, 2, pvalue, sizeof(pvalue));
+        if (res < 0) {
+            SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_copy_substring failed");
+            goto error;
+        }
+        value = DetectMedatataHashAdd(de_ctx, pvalue);
+        if (value == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "can't create metadata value");
+            goto error;
+        }
+
+        SCLogDebug("key: %s, value: %s", key, value);
+
+        dkv = SCMalloc(sizeof(DetectMetadata));
+        if (dkv) {
+            dkv->key = key;
+            if (!dkv->key) {
+                SCFree(dkv);
+            } else {
+                dkv->value = value;
+                if (!dkv->value) {
+                    SCFree((void *)dkv->key);
+                    SCFree(dkv);
+                } else {
+                    dkv->next = s->metadata;
+                    s->metadata = dkv;
+                }
+            }
+        } else {
+            goto error;
+        }
+
+        kv = strtok_r(NULL, ",", &saveptr);
+    } while (kv);
+
+    return 0;
+
+error:
+    if (key)
+        SCFree((void *)key);
+    if (value)
+        SCFree((void *)value);
+    return -1;
+}
+
+static int DetectMetadataSetup(DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    if (DetectEngineMustParseMetadata()) {
+        DetectMetadataParse(de_ctx, s, rawstr);
+    }
+
+    return 0;
+}
+
+#ifdef UNITTESTS
+
+static int DetectMetadataParseTest01(void)
+{
+    DetectEngineCtx *de_ctx = DetectEngineCtxInit();
+    FAIL_IF_NULL(de_ctx);
+
+    Signature *sig = DetectEngineAppendSig(de_ctx, "alert tcp any any -> any any (metadata: toto 1; sid:1; rev:1;)");
+    FAIL_IF_NULL(sig);
+    FAIL_IF(sig->metadata); 
+
+    DetectEngineCtxFree(de_ctx);
+    PASS;
+}
+
+static int DetectMetadataParseTest02(void)
+{
+    DetectEngineCtx *de_ctx = NULL;
+    DetectMetadata *dm;
+
+    DetectEngineSetParseMetadata();
+
+    de_ctx = DetectEngineCtxInit();
+    FAIL_IF_NULL(de_ctx);
+    Signature *sig = DetectEngineAppendSig(de_ctx, "alert tcp any any -> any any (metadata: toto 1; metadata: titi 2, jaivu gros_minet; sid:1; rev:1;)");
+    FAIL_IF_NULL(sig);
+    FAIL_IF_NULL(sig->metadata); 
+    FAIL_IF_NULL(sig->metadata->key); 
+    FAIL_IF(strcmp("jaivu", sig->metadata->key));
+    FAIL_IF(strcmp("gros_minet", sig->metadata->value));
+    FAIL_IF_NULL(sig->metadata->next); 
+    dm = sig->metadata->next;
+    FAIL_IF(strcmp("titi", dm->key));
+    dm = dm->next;
+    FAIL_IF_NULL(dm);
+    FAIL_IF(strcmp("toto", dm->key));
+
+    DetectEngineCtxFree(de_ctx);
+    PASS;
```

Comment:
should DetectEngineSetParseMetadata be reset here? Ideally a test doesn't modify global state

Reply:
not permanently I mean